File: backend/app/api/v1/routes/academic.py
"""
app/api/v1/routes/academic.py

Cascading hierarchy routes for academic onboarding and assignments.
"""

import uuid
from typing import List

from fastapi import APIRouter, Depends, Query
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.db.models.academic import Institution, Campus, College, Department, Option, ClassGroup, ClassSection, Course, AcademicPeriod
from app.db.schemas.academic import (
    InstitutionResponse,
    CampusResponse,
    CollegeResponse,
    DepartmentResponse,
    OptionResponse,
    ClassGroupResponse,
    ClassSectionResponse,
    CourseResponse,
    AcademicPeriodResponse,
)
from app.db.session import get_db

router = APIRouter(prefix="/academic-hierarchy", tags=["Academic Hierarchy"])


@router.get("/levels", response_model=List[int])
async def get_levels(
    institution_id: uuid.UUID | None = None,
    db: AsyncSession = Depends(get_db)
):
    """List all distinct academic levels defined in the system or institution."""
    from sqlalchemy import func
    stmt = select(func.distinct(ClassGroup.level)).where(ClassGroup.level != None)
    if institution_id:
        # Join through Option and Department to filter by institution
        stmt = stmt.join(Option).join(Department).where(Department.institution_id == institution_id)
        
    result = await db.execute(stmt)
    # Sort levels numerically
    levels = sorted([row for row in result.scalars().all()])
    return levels


@router.get("/institutions", response_model=List[InstitutionResponse])
async def get_institutions(db: AsyncSession = Depends(get_db)):
    """List all active institutions."""
    stmt = select(Institution).where(Institution.is_active == True)
    result = await db.execute(stmt)
    return result.scalars().all()


@router.get("/campuses", response_model=List[CampusResponse])
async def get_campuses(
    institution_id: uuid.UUID, 
    db: AsyncSession = Depends(get_db)
):
    """List campuses within an institution."""
    stmt = select(Campus).where(
        Campus.institution_id == institution_id,
        Campus.is_active == True
    )
    result = await db.execute(stmt)
    return result.scalars().all()


@router.get("/colleges", response_model=List[CollegeResponse])
async def get_colleges(
    institution_id: uuid.UUID | None = None,
    campus_id: uuid.UUID | None = None, 
    db: AsyncSession = Depends(get_db)
):
    """List colleges within a campus or institution."""
    stmt = select(College).where(College.is_active == True)
    if campus_id:
        stmt = stmt.where(College.campus_id == campus_id)
    elif institution_id:
        stmt = stmt.where(College.institution_id == institution_id)
    else:
        return []
    result = await db.execute(stmt)
    return result.scalars().all()


@router.get("/departments", response_model=List[DepartmentResponse])
async def get_departments(
    institution_id: uuid.UUID | None = None,
    campus_id: uuid.UUID | None = None,
    college_id: uuid.UUID | None = None,
    db: AsyncSession = Depends(get_db)
):
    """List departments within an institution, campus, or college."""
    stmt = select(Department).where(Department.is_active == True)
    if college_id:
        stmt = stmt.where(Department.college_id == college_id)
    elif campus_id:
        stmt = stmt.where(Department.campus_id == campus_id)
    elif institution_id:
        stmt = stmt.where(Department.institution_id == institution_id)
    else:
        return []

    result = await db.execute(stmt)
    return result.scalars().all()


@router.get("/options", response_model=List[OptionResponse])
async def get_options(
    department_id: uuid.UUID, 
    db: AsyncSession = Depends(get_db)
):
    """List academic options/programs within a department."""
    stmt = select(Option).where(
        Option.department_id == department_id,
        Option.is_active == True
    )
    result = await db.execute(stmt)
    return result.scalars().all()


@router.get("/class-groups", response_model=List[ClassGroupResponse])
async def get_class_groups(
    option_id: uuid.UUID, 
    db: AsyncSession = Depends(get_db)
):
    """List class groups (levels) within an option."""
    stmt = select(ClassGroup).where(
        ClassGroup.option_id == option_id,
        ClassGroup.is_active == True
    )
    result = await db.execute(stmt)
    return result.scalars().all()


@router.get("/sections", response_model=List[ClassSectionResponse])
async def get_sections(
    class_group_id: uuid.UUID | None = None,
    department_id: uuid.UUID | None = None,
    course_id: uuid.UUID | None = None,
    db: AsyncSession = Depends(get_db)
):
    """List specific sections within a class group, department, or for a specific course."""
    stmt = select(ClassSection).where(ClassSection.is_active == True)
    if class_group_id:
        stmt = stmt.where(ClassSection.class_group_id == class_group_id)
    if department_id:
        stmt = stmt.where(ClassSection.department_id == department_id)
    if course_id:
        from app.db.models.academic import TeachingAssignment
        stmt = stmt.join(TeachingAssignment, TeachingAssignment.class_section_id == ClassSection.id).where(
            TeachingAssignment.course_id == course_id
        ).distinct()

    if not class_group_id and not department_id and not course_id:
        return []

    result = await db.execute(stmt)
    return result.scalars().all()


@router.get("/courses", response_model=List[CourseResponse])
async def get_courses(
    department_id: uuid.UUID, 
    db: AsyncSession = Depends(get_db)
):
    """List modules/courses within a department."""
    from app.db.models.academic import CourseDepartment
    
    stmt = (
        select(Course)
        .join(CourseDepartment, CourseDepartment.course_id == Course.id)
        .where(
            CourseDepartment.department_id == department_id,
            Course.is_active == True
        )
    )
    result = await db.execute(stmt)
    return result.scalars().all()


@router.get("/academic-periods", response_model=List[AcademicPeriodResponse])
async def get_periods(
    institution_id: uuid.UUID | None = None, 
    db: AsyncSession = Depends(get_db)
):
    """List academic periods (semesters) within an institution, or all if none provided."""
    stmt = select(AcademicPeriod).where(AcademicPeriod.is_active == True)
    if institution_id:
        stmt = stmt.where(AcademicPeriod.institution_id == institution_id)
        
    result = await db.execute(stmt)
    return result.scalars().all()

File: frontend/lib/api/academic.ts
import { apiClient } from "./client";
import { 
  InstitutionResponse, 
  CampusResponse, 
  CollegeResponse, 
  DepartmentResponse, 
  OptionResponse, 
  ClassGroupResponse, 
  ClassSectionResponse,
  CourseResponse,
  AcademicPeriodResponse
} from "./academic-types";

export type AcademicInstitution = InstitutionResponse
export type AcademicCampus = CampusResponse
export type AcademicCollege = CollegeResponse
export type AcademicDepartment = DepartmentResponse
export type AcademicOption = OptionResponse
export type AcademicClassGroup = ClassGroupResponse
export type AcademicClassSection = ClassSectionResponse
export type AcademicCourse = CourseResponse
export type AcademicPeriod = AcademicPeriodResponse

export interface TeachingAssignment {
  id: string;
  lecturer_id: string;
  institution_id: string;
  campus_id?: string;
  college_id?: string;
  department_id: string;
  option_id?: string;
  course_id: string;
  class_section_id?: string;
  academic_period_id: string;
  role: "MAIN_LECTURER" | "ASSISTANT_LECTURER" | "SUPERVISOR" | "REVIEWER";
  is_active: boolean;
  academic_year?: string;
  institution_name?: string;
  campus_name?: string;
  college_name?: string;
  department_name?: string;
  option_name?: string;
  course_name?: string;
  course_code?: string;
  class_section_name?: string;
  class_group_name?: string;
  class_group_level?: number;
}

export const academicApi = {
  getLevels: async (institutionId?: string): Promise<number[]> => {
    const url = institutionId 
      ? `/academic-hierarchy/levels?institution_id=${institutionId}` 
      : "/academic-hierarchy/levels";
    return apiClient(url);
  },
  getInstitutions: async (): Promise<AcademicInstitution[]> => {
    return apiClient("/academic-hierarchy/institutions");
  },
  getCampuses: async (institutionId: string): Promise<AcademicCampus[]> => {
    return apiClient(`/academic-hierarchy/campuses?institution_id=${institutionId}`);
  },
  getColleges: async (params: { institution_id?: string; campus_id?: string }): Promise<AcademicCollege[]> => {
    const query = new URLSearchParams();
    if (params.institution_id) query.append("institution_id", params.institution_id);
    if (params.campus_id) query.append("campus_id", params.campus_id);
    return apiClient(`/academic-hierarchy/colleges?${query.toString()}`);
  },
  getDepartments: async (params: { institution_id?: string; campus_id?: string; college_id?: string }): Promise<AcademicDepartment[]> => {
    const query = new URLSearchParams();
    if (params.institution_id) query.append("institution_id", params.institution_id);
    if (params.campus_id) query.append("campus_id", params.campus_id);
    if (params.college_id) query.append("college_id", params.college_id);
    return apiClient(`/academic-hierarchy/departments?${query.toString()}`);
  },
  getOptions: async (departmentId: string): Promise<AcademicOption[]> => {
    return apiClient(`/academic-hierarchy/options?department_id=${departmentId}`);
  },
  getClassGroups: async (optionId: string): Promise<AcademicClassGroup[]> => {
    return apiClient(`/academic-hierarchy/class-groups?option_id=${optionId}`);
  },
  getSections: async (params: { class_group_id?: string; department_id?: string } | string): Promise<AcademicClassSection[]> => {
    if (typeof params === "string") {
      return apiClient(`/academic-hierarchy/sections?class_group_id=${params}`);
    }
    const query = new URLSearchParams();
    if (params.class_group_id) query.append("class_group_id", params.class_group_id);
    if (params.department_id) query.append("department_id", params.department_id);
    return apiClient(`/academic-hierarchy/sections?${query.toString()}`);
  },
  getCourses: async (departmentId: string): Promise<AcademicCourse[]> => {
    return apiClient(`/academic-hierarchy/courses?department_id=${departmentId}`);
  },
  getPeriods: async (institutionId?: string): Promise<AcademicPeriod[]> => {
    const url = institutionId 
      ? `/academic-hierarchy/academic-periods?institution_id=${institutionId}`
      : "/academic-hierarchy/academic-periods";
    return apiClient(url);
  },
  getMyAssignments: async (): Promise<TeachingAssignment[]> => {
    return apiClient("/lecturers/me/assignments");
  },
};

// Also export as individual functions for robustness
export const getAcademicPeriods = academicApi.getPeriods;

export const adminAcademicApi = {
  // Teaching Assignments
  assignLecturer: async (data: any) => {
    return apiClient("/admin/academic/assignments", {
      method: "POST",
      body: JSON.stringify(data),
    });
  },
  getLecturerAssignments: async (lecturerId?: string): Promise<TeachingAssignment[]> => {
    const url = lecturerId 
      ? `/admin/academic/assignments?lecturer_id=${lecturerId}`
      : "/admin/academic/assignments";
    return apiClient(url);
  },
  removeAssignment: async (assignmentId: string) => {
    return apiClient(`/admin/academic/assignments/${assignmentId}`, {
      method: "DELETE",
    });
  },

  // Academic Structure Creation
  createCampus: async (data: any) => {
    return apiClient("/admin/academic/campuses", {
      method: "POST",
      body: JSON.stringify(data),
    });
  },
  createCollege: async (data: any) => {
    return apiClient("/admin/academic/colleges", {
      method: "POST",
      body: JSON.stringify(data),
    });
  },
  createDepartment: async (data: any) => {
    return apiClient("/admin/academic/departments", {
      method: "POST",
      body: JSON.stringify(data),
    });
  },
  createOption: async (data: any) => {
    return apiClient("/admin/academic/options", {
      method: "POST",
      body: JSON.stringify(data),
    });
  },
  createClassGroup: async (data: any) => {
    return apiClient("/admin/academic/class-groups", {
      method: "POST",
      body: JSON.stringify(data),
    });
  },
  createSection: async (data: any) => {
    return apiClient("/admin/academic/sections", {
      method: "POST",
      body: JSON.stringify(data),
    });
  },
  getAcademicPeriods: async (institutionId?: string): Promise<AcademicPeriod[]> => {
    const url = institutionId 
      ? `/admin/academic/academic-periods?institution_id=${institutionId}`
      : "/admin/academic/academic-periods";
    return apiClient(url);
  },
  createAcademicPeriod: async (data: any) => {
    return apiClient("/admin/academic/academic-periods", {
      method: "POST",
      body: JSON.stringify(data),
    });
  },
  createCourse: async (data: any) => {
    return apiClient("/admin/academic/courses", {
      method: "POST",
      body: JSON.stringify(data),
    });
  },
  updateEntity: async (entityType: string, id: string, data: any) => {
    return apiClient(`/admin/academic/${entityType}/${id}`, {
      method: "PATCH",
      body: JSON.stringify(data),
    });
  },
};
